_CS141/part2.md

```diff
@@ -0,0 +1,308 @@
+
+Here the definitions we use are.
+
+```haskell
+length :: [a] -> Int
+length []     = 0
+length (x:xs) = 1 + length xs
+
+take :: Int -> [a] -> [a]
+take 0 _     = []
+take n []    = []
+take n (x:xs) = x : take (n-1) xs
+
+map :: (a -> b) -> [a] -> [b]
+map f []     = []
+map f (x:xs) = f x : map f xs
+```
+
+**Answer.**
+
+```haskell
+   length (take 2 (map even [1,2,3,4]))
+=> length (take 2 (even 1 : map even [2,3,4]))
+=> length (even 1 : take (2-1) (map even [2,3,4]))
+=> 1 + length (take (2-1) (map even [2,3,4]))
+=> 1 + length (take 1 (map even [2,3,4]))
+=> 1 + length (take 1 (even 2 : map even [3,4]))
+=> 1 + length (even 2 : take (1-1) (map even [3,4]))
+=> 1 + (1 + length (take (1-1) (map even [3,4])))
+=> 1 + (1 + length (take 0 (map even [3,4]))) 
+=> 1 + (1 + length [])
+=> 1 + (1 + 0)
+=> 1 + 1
+=> 2
+```
+
+Questions may ask about call-by-value and call-by-name too, so know this topic well.
+
+## Closures
+
+> A structure in memory at runtime that represents a function and its environments (i.e scope).
+
+We can think of it as an array of pointers 
+
+- 1st pointer points to some code
+- other pointers point to other closures.
+
+**Example.**
+
+```haskell
+not :: Bool -> Bool
+not True = False
+not False = True
+```
+
+In memory, you have a particular memory location/address that stores a pointer to the code (we don’t have to worry about where the code is.)
+
+- Every usage of `not` in the code is a pointer to this memory location.
+
+> Functions in Haskell (and other functional programming languages) are first class values, which means they can be returned by other functions or given to functions as arguments (higher-order-functions).
+
+**Example.**
+
+```haskell
+f :: Int -> a -> Int
+f x = let g y = x + 5
+      in g
+```
+
+In memory, we have a static closure of `f` as mentioned above for `not`
+
+- Initially when the program is started, there is no closure for `g`. Only have closures for top-level definitions like `f` here when program starts.
+- When we start evaluating `f`, every call to `f` will dynamically allocate a closure to `g` on the heap (this is only for that particular invocation of `f`)
+  - This closure is comprised of a pointer to code for `g` and a pointer to `x`
+  - The reason for this is because the body of `g` refers to `x`, which is a parameter of `f`.
+  - Because `g` is defined within the scope of `f`, it has access to whatever is in scope of `f`, which `x` is.
+- It doesn’t matter if the stack frame for `f` is removed because `g` still has a reference to what `x` **was**.
+
+**Example.**
+
+```haskell
+length (take 2 (map even [1,2,3,4])) 
+-- is translated into
+let zs4 = 4 : []
+    zs3 = 3 : zs4
+    zs2 = 2 : zs3
+    zs  = 1 : zs2
+    ys  = map even zs
+    xs  = take 2 ys
+in length xs
+-- by the compiler
+```
+
+When this evaluation gets evaluated at runtime, a dynamically allocated closure for each of these “variables” (so zs4, zs3, … , xs) is made.
```

Comment:
I think this should say "expression" instead of "evaluation" (but I can't suggest it since the PR has already been merged). "[..], a closure is dynamically allocated for each of these [..]" might be a better wording to avoid the "is made" at the end as well